File: python/forge3d/_degradation.py
# python/forge3d/_degradation.py
# CENSOR Task 10: pure-Python degradation sink, mirroring the Rust one in
# src/core/degradation.rs. Every Python-side fallback / placeholder /
# absent-capability path records a structured entry here; the RenderCertificate
# (forge3d.diagnostics.render_certificate) drains it and merges it with the
# native sink.
# RELEVANT FILES: src/core/degradation.rs, python/forge3d/diagnostics.py,
# python/forge3d/certificate.py

"""Process-global degradation sink (Python side).

Byte-for-byte behavioural twin of the Rust sink in
``src/core/degradation.rs``: entries are deduplicated on ``(kind, name)`` so
per-frame fallbacks do not flood the certificate, and :func:`snapshot` returns
them sorted by ``(kind, name)``.
"""

from __future__ import annotations

import threading
from typing import Dict, List, Tuple

# (kind, name) -> consequence. First writer for a given (kind, name) wins, so
# the recorded consequence is stable regardless of how many times a per-frame
# fallback fires.
_SINK: Dict[Tuple[str, str], str] = {}
_ACTIVE_CAPTURE: Dict[Tuple[str, str], str] | None = None
_LAST_CAPTURE: Dict[Tuple[str, str], str] = {}
_LOCK = threading.Lock()


def record(kind: str, name: str, consequence: str) -> None:
    """Record a degradation, deduplicated on ``(kind, name)``.

    The first recorded ``consequence`` for a given ``(kind, name)`` is kept;
    later calls with the same key are ignored (mirrors the Rust sink).
    """
    key = (str(kind), str(name))
    global _ACTIVE_CAPTURE
    with _LOCK:
        if key not in _SINK:
            _SINK[key] = str(consequence)
        if _ACTIVE_CAPTURE is not None and key not in _ACTIVE_CAPTURE:
            _ACTIVE_CAPTURE[key] = str(consequence)


def snapshot() -> List[Dict[str, str]]:
    """Return all recorded degradations as dicts, sorted by ``(kind, name)``."""
    with _LOCK:
        items = sorted(_SINK.items())
    return [
        {"kind": kind, "name": name, "consequence": consequence}
        for (kind, name), consequence in items
    ]


def begin_capture() -> None:
    """Start a fresh render-local capture."""
    global _ACTIVE_CAPTURE
    with _LOCK:
        _ACTIVE_CAPTURE = {}


def finish_capture() -> None:
    """Freeze the active render-local capture for certificate assembly."""
    global _ACTIVE_CAPTURE, _LAST_CAPTURE
    with _LOCK:
        _LAST_CAPTURE = dict(_ACTIVE_CAPTURE or {})
        _ACTIVE_CAPTURE = None


def capture_snapshot() -> List[Dict[str, str]]:
    """Return the last completed render's Python degradations."""
    with _LOCK:
        items = sorted(_LAST_CAPTURE.items())
    return [
        {"kind": kind, "name": name, "consequence": consequence}
        for (kind, name), consequence in items
    ]


def abort_capture() -> None:
    """Discard an incomplete render-local capture."""
    global _ACTIVE_CAPTURE
    with _LOCK:
        _ACTIVE_CAPTURE = None


def clear() -> None:
    """Reset the sink. Exposed so tests and callers can isolate renders."""
    global _ACTIVE_CAPTURE, _LAST_CAPTURE
    with _LOCK:
        _SINK.clear()
        _ACTIVE_CAPTURE = None
        _LAST_CAPTURE = {}

File: src/core/degradation.rs
//! Global degradation sink: every fallback/placeholder/absent-capability path
//! records a structured entry here; the RenderCertificate drains it.
use serde::Serialize;
use std::cell::RefCell;
use std::sync::Mutex;

#[derive(Clone, Debug, PartialEq, Eq, Serialize)]
pub struct Degradation {
    pub kind: String,
    pub name: String,
    pub consequence: String,
}

static SINK: Mutex<Vec<Degradation>> = Mutex::new(Vec::new());

#[cfg(test)]
pub(crate) static TEST_SINK_LOCK: Mutex<()> = Mutex::new(());

thread_local! {
    static CAPTURE: RefCell<Option<Vec<Degradation>>> = const { RefCell::new(None) };
}

/// Record a degradation. Deduplicated on (kind, name) so per-frame fallbacks
/// do not flood the certificate.
pub fn record_degradation(kind: &str, name: &str, consequence: &str) {
    let entry = Degradation {
        kind: kind.to_string(),
        name: name.to_string(),
        consequence: consequence.to_string(),
    };
    let mut sink = SINK.lock().unwrap_or_else(|p| p.into_inner());
    if !sink.iter().any(|d| d.kind == kind && d.name == name) {
        sink.push(entry.clone());
    }
    drop(sink);
    CAPTURE.with(|capture| {
        if let Some(captured) = capture.borrow_mut().as_mut() {
            if !captured.iter().any(|d| d.kind == kind && d.name == name) {
                captured.push(entry);
            }
        }
    });
}

pub fn degradations_snapshot() -> Vec<Degradation> {
    SINK.lock().unwrap_or_else(|p| p.into_inner()).clone()
}

pub fn begin_degradation_capture() {
    CAPTURE.with(|capture| *capture.borrow_mut() = Some(Vec::new()));
}

pub fn finish_degradation_capture() -> Vec<Degradation> {
    CAPTURE.with(|capture| capture.borrow_mut().take().unwrap_or_default())
}

pub fn abort_degradation_capture() {
    CAPTURE.with(|capture| {
        capture.borrow_mut().take();
    });
}

/// Reset the sink. Exposed to Python so tests can isolate renders.
pub fn clear_degradations() {
    SINK.lock().unwrap_or_else(|p| p.into_inner()).clear();
}

#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn dedup_on_kind_and_name() {
        let _lock = TEST_SINK_LOCK.lock().unwrap_or_else(|p| p.into_inner());
        clear_degradations();
        record_degradation("capability_absent", "timestamp_query", "a");
        record_degradation("capability_absent", "timestamp_query", "b");
        record_degradation("cpu_fallback", "timestamp_query", "c");
        assert_eq!(degradations_snapshot().len(), 2);
        clear_degradations();
    }

    #[test]
    fn degradation_capture_is_render_local() {
        let _lock = TEST_SINK_LOCK.lock().unwrap_or_else(|p| p.into_inner());
        clear_degradations();
        record_degradation("pre_render", "excluded", "before");
        record_degradation("repeated", "included", "before");

        begin_degradation_capture();
        record_degradation("repeated", "included", "during");
        record_degradation("in_render", "included", "during");
        let captured = finish_degradation_capture();

        assert!(!captured.iter().any(|d| d.name == "excluded"));
        assert!(captured
            .iter()
            .any(|d| d.kind == "repeated" && d.name == "included"));
        assert!(captured.iter().any(|d| d.kind == "in_render"));
        clear_degradations();
    }
}
